contracts/serverless-v2/CommonChainExecutors.sol

```diff
@@ -346,48 +347,42 @@ contract CommonChainExecutors is
     function _updateOnSubmitOutput(
         address _executorKey
     ) internal {
-        // add back the node to the tree as now it can accept a new job
-        if(executors[_executorKey].status && executors[_executorKey].activeJobs == executors[_executorKey].jobCapacity)
-            _insert_unchecked(_executorKey, uint64(executors[_executorKey].stakeAmount));
-
-        executors[_executorKey].activeJobs -= 1;
-
-        // if user has initiated unstake then release tokens only if no jobs are pending
-        if(executors[_executorKey].unstakeStatus && executors[_executorKey].activeJobs == 0) {
-            uint256 amount = executors[_executorKey].stakeAmount < executors[_executorKey].unstakeAmount ? executors[_executorKey].stakeAmount : executors[_executorKey].unstakeAmount;
-            executors[_executorKey].stakeAmount -= amount;
-            TOKEN.safeTransfer(_msgSender(), amount);
-            executors[_executorKey].unstakeAmount = 0;
-            // update in tree only if the user has not initiated deregistration
-            if(executors[_executorKey].status)
-                _update_unchecked(_executorKey, uint64(executors[_executorKey].stakeAmount));
-        }
-
-        if(!executors[_executorKey].status && executors[_executorKey].activeJobs == 0) {
-            delete executors[_executorKey];
-            _revokeEnclaveKey(_executorKey);
-        }
+        _preUpdate(_executorKey);
+        _postUpdate(_executorKey);
     }
 
     function _updateOnExecutionTimeoutSlash(
         address _executorKey,
         bool _hasExecutedJob
+    ) internal {
+        _preUpdate(_executorKey);
+
+        // TODO: slash executor if failed to perform the job
+        if(!_hasExecutedJob) {}
```

Comment:
slashing should be performed first, so that stakes are updated and then call the preUpdate which will add the node with new stakes to the tree. And this way you can merge preUpdate and postUpdate into one function.